File: Compass/Agents/Codex.cs
using System.Diagnostics;

namespace Compass.Agents;

public class Codex : Agent
{
    public override string Name => "Codex";

    public override async Task EnsureLogin()
    {
        var checkLogin = await ProcessUtils.Run(
            Environment.CurrentDirectory,
            "codex",
            "login status");

        if (checkLogin.ExitCode != 0)
        {
            var loginProcess = Process.Start(new ProcessStartInfo
            {
                FileName = "codex",
                Arguments = "login --device-auth",
                WorkingDirectory = Environment.CurrentDirectory,
                RedirectStandardOutput = false,
                RedirectStandardError = false,
                UseShellExecute = false
            }) ?? throw new InvalidOperationException("Failed to start Codex login.");

            await loginProcess.WaitForExitAsync();

            if (loginProcess.ExitCode != 0)
            {
                throw new Exception("Codex login failed.");
            }
        }
    }

    public override async Task<AgentOutput> Execute(string prompt, string model, string workingDirectory)
    {
        var processOutput = await ProcessUtils.Run(
            workingDirectory,
            "codex",
            $"exec --model {model.EscapeArg()} --sandbox danger-full-access {prompt.EscapeArg()}");

        string? agentOutput = processOutput.StdOut;

        Logger.Log($"Collecting git diff after agent execution", Logger.LogLevel.Verbose);

        var diff = await ProcessUtils.Git(workingDirectory, "--no-pager diff");

        return new AgentOutput()
        {
            StdOut = agentOutput,
            StdErr = processOutput.StdErr,
            GitDiff = diff.StdOut
        };
    }
}

File: src/agents/codex.ts
import type { AgentOutput } from "../models";
import { run } from "../utils";
import { logger } from "../logging";
import { Agent, type AgentOptions } from "./agent";

export class Codex extends Agent {
    constructor(options: AgentOptions) {
        super("Codex", options);
    }

    override async execute(prompt: string, model: string, workingDirectory: string): Promise<AgentOutput> {
        logger.trace(`Executing Codex with model ${model} on prompt ${prompt}`);
        
        const sandboxParameters = this.options.allowFullAccess ? ["--sandbox", "danger-full-access"] : [];
        const processOutput = await run(
            workingDirectory,
            "codex",
            "exec",
            "--model", model, 
            ...sandboxParameters,
            prompt);

        logger.trace("Codex stdOut: " + processOutput.stdOut);
        logger.trace("Codex stdErr: " + processOutput.stdErr);
        logger.trace("Codex exitCode: " + processOutput.exitCode);

        logger.trace("Collecting git diff after agent execution");
        
        const diff = await run(workingDirectory, "git", "--no-pager", "diff");

        return {
            stdOut: Bun.stripANSI(processOutput.stdOut.trim()),
            stdErr: Bun.stripANSI(processOutput.stdErr.trim()),
            exitCode: processOutput.exitCode,
            gitDiff: diff.stdOut.trim()
        };
    }

    override async init(): Promise<void> {
        const checkLogin = await run(".", "codex", "login", "status");

        if (checkLogin.exitCode !== 0) {
            const loginProcess = Bun.spawn(
                ["codex", "login", "--device-auth"], 
                { 
                    stdin: "inherit", 
                    stdout: "inherit", 
                    stderr: "inherit" 
                }
            );
            
            await loginProcess.exited;

            if (loginProcess.exitCode !== 0) {
                throw new Error("Failed to log in to Codex CLI.");
            }
             
        }
    }
    
}